jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -507,15 +515,28 @@ public String toString()
         // hi: positive=open/maxUsage counter; negative=closed; MIN_VALUE pending
         // lo: multiplexing counter
         private final AtomicBiInteger state;
-
         // The pooled item.  This is not volatile as it is set once and then never changed.
         // Other threads accessing must check the state field above first, so a good before/after
         // relationship exists to make a memory barrier.
         private T pooled;
+        private volatile int maxMultiplex;
```

Comment:
I'm still wondering if we should store 2 values here:
```suggestion
        private volatile int remoteMaxMultiplex;
        private volatile int localMaxMultiplex;
```

We set localMaxMultiplex when we send a settings frame, we set remoteMaxMultiplex when we receive a settings frame.   If we change settings, local changes first and then remote changes when we get the ack, otherwise remote changes first and local changes second.

It will always be a hard error to have current greater than the localMaxMultiplex. 

If the localMaxMultiplex is less than the remoteMaxMultiplex then that doesn't matter, it is still and error have current greater than the local.

If the localMaxMultiplex is greater than the remoteMaxMultiplex then we have to be more nuanced.  It is an error for an remote created stream to exceed the remoteMaxMultiplex, but a locally created stream is allowed to let current exceed remoteMaxMultiplex but not localMaxMultiplex.

I think we really need to deal with this type of dynamic max streams adjustment. We have already seen one example in the wild... others will follow even if it is a bad idea.

Reply:
Also this kind of exceeds my threshold for unused features in a base utility class.  Perhaps we need to split `Pool` into `Pool` and `MultiplexPool`?

Reply:
> We set localMaxMultiplex when we send a settings frame

No, we don't. I don't think what you say is correct, although I understand the underlying idea of handling the case where a second incoming SETTINGS frame has a smaller `max_concurrent_streams` value.
However, I would handle this latter case in a different PR.

Reply:
Let's pair program on this tonight. I think it can be fixed in this PR.

Reply:
@sbordet  As we discussed on hangout, we don't need to fix all the problems in this PR, but we should not add wrong things to the Pool.   We should externalize the multiplexing checks as per my draft PR

Reply:
@gregw I second this goal. I'd be very happy if we managed to externalize the multiplexing out of the low-level `Pool` as IMHO it was a design mistake.

That design also has adverse effects on performance as the delegation of the multiplexing accounting is delegated to a lower-level that the only thing it can do is to track it by using an atomic counter that's naturally going to be highly contended, i.e.: the HTTP2 pool is much slower when connections are multiplexed than when they're not because there's a ton of contention on the CAS loops around the multiplexing counters. If the multiplexing counter was tracked at the connection pool layer, we could add some contention-mitigating strategies.